src/strategy/exponential_backoff.rs

```diff
@@ -1,46 +1,115 @@
 use std::time::Duration;
 use std::iter::Iterator;
-use std::u64::{MAX as U64_MAX};
+use std::u64::MAX as U64_MAX;
+use strategy::MaxDelay;
 
 /// A retry strategy driven by exponential back-off.
 ///
 /// The power corresponds to the number of past attempts.
-#[derive(Clone)]
+#[derive(Debug, Clone)]
 pub struct ExponentialBackoff {
     current: u64,
-    base: u64
+    base: u64,
+    unit: Unit,
+    max_delay: Option<MaxDelay>,
+}
+
+#[derive(Debug, Copy, Clone)]
+enum Unit {
+    Millis,
+    Secs,
+}
+
+impl Unit {
+    /// Make a new `Duration` from the `Unit` variant using the provided base
+    fn new_duration(&self, base: u64) -> Duration {
+        match *self {
+            Unit::Millis => Duration::from_millis(base),
+            Unit::Secs => Duration::from_secs(base),
+        }
+    }
 }
 
 impl ExponentialBackoff {
     /// Constructs a new exponential back-off strategy,
     /// given a base duration in milliseconds.
     ///
+    /// An optional maximum delay can be specified, ensuring no retry delay exceeds the provided
+    /// `Duration`.
+    ///
     /// The resulting duration is calculated by taking the base to the `n`-th power,
     /// where `n` denotes the number of past attempts.
-    pub fn from_millis(base: u64) -> ExponentialBackoff {
-        ExponentialBackoff{current: base, base: base}
+    pub fn from_millis(base: u64, max_delay: Option<Duration>) -> ExponentialBackoff {
+        ExponentialBackoff {
+            current: base,
+            base: base,
+            unit: Unit::Millis,
+            max_delay: MaxDelay::new(max_delay),
+        }
+    }
+
+    /// Constructs a new exponential back-off strategy,
+    /// given a base duration in seconds.
+    ///
+    /// An optional maximum delay can be specified, ensuring no retry delay exceeds the provided
+    /// `Duration`.
+    ///
+    /// The resulting duration is calculated by taking the base to the `n`-th power,
+    /// where `n` denotes the number of past attempts.
+    pub fn from_secs(base: u64, max_delay: Option<Duration>) -> ExponentialBackoff {
+        ExponentialBackoff {
+            current: base,
+            base: base,
+            unit: Unit::Secs,
+            max_delay: MaxDelay::new(max_delay),
+        }
     }
 }
 
 impl Iterator for ExponentialBackoff {
     type Item = Duration;
 
     fn next(&mut self) -> Option<Duration> {
-        let duration = Duration::from_millis(self.current);
-
-        if let Some(next) = self.current.checked_mul(self.base) {
-            self.current = next;
+        let duration = if let Some(ref mut max_delay) = self.max_delay {
```

Comment:
To separate the concerns, it seems like this logic could easily live on `MaxDelay`:

```
impl MaxDelay {
    pub fn apply(duration: Duration) -> Duration {
        ...
    }
}
```